docs/proposals/api-group.md

```diff
@@ -0,0 +1,135 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/api-group.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Supporting multiple API groups
+
+## Goal
+
+1. Breaking the monolithic v1 API into modular groups
+
+2. Supporting different versions in different groups
+
+3. Supporting identically named kinds to exist in different groups
+
+4. Enabling/disabling groups individually
+
+5. Allowing client to discover the API groups supported by the server
+
+6. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
+
+7. Allowing users to omit group name when using kubectl
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  For fast implementation, we will not add a new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
```

Comment:
Customers have a standard proxy somewhere (an authenticating reverse http
proxy) at mycompany.com.  When they deploy, they prefer to front Kube with
their proxy - but because it's shared, the root namespace can't be
claimed.  So they might want to host at www.mycompany.com/kube/*.  To
login, a client might specify www.mycompany.com/kube instead of just
www.mycompany.com

It's less common these days, but it comes up in large companies
occasionally.  When we design logins and clients we just try not to assume
that we get to dictate the API space.  So /apis as a convention is fine,
but the client library and config today supports passing a URL to the
initial host which might provide the prefix.  There should be very little
code after that initial point that requires knowing the prefix - that's why
would do API discovery.

On Sep 2, 2015, at 9:46 PM, Chao Xu <notifications@github.com> wrote:

In docs/proposals/api-group.md
<https://github.com/kubernetes/kubernetes/pull/12951#discussion_r38606105>:

> +
> +2. conversion.Scheme:
> +
> +  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
> +
> +3. meta.RESTMapper:
> +
> +  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
> +
> +  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
> +
> +## Server-side implementation
> +
> +1. resource handlers' URL:
> +
> +  We will force the URL to be in the form of prefix/group/version/...

We get asked all the time to support hosting urls under someone else's
proxy at a non root path. So practically I'd like to make apis the
convention and not force clients into assuming the base URL has no path.

@smarterclayton <https://github.com/smarterclayton> , could you explain
this in other words? I found it hard to follow. Thank you.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12951/files#r38606105>.